core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceRequesterPaysComponent.scala

```diff
@@ -45,13 +46,30 @@ trait WorkspaceRequesterPaysComponent {
       query.exists.result
     }
 
-    def listAllForUser(workspaceName: WorkspaceName, userEmail: RawlsUserEmail): ReadAction[Seq[String]] = {
-      existingRecordsForUserQuery(workspaceName, userEmail).map(_.serviceAccountEmail).result
+    def listAllForUser(workspaceName: WorkspaceName, userEmail: RawlsUserEmail): ReadAction[Seq[BondServiceAccountEmail]] = {
+      existingRecordsForUserQuery(workspaceName, userEmail).map(_.serviceAccountEmail).result.map(_.map(BondServiceAccountEmail))
+    }
+
+    def listAllForWorkspace(workspaceName: WorkspaceName): ReadAction[Map[RawlsUserEmail, Seq[BondServiceAccountEmail]]] = {
+      existingRecordsForWorkspaceQuery(workspaceName).map(rec => rec.userEmail -> rec.serviceAccountEmail).result
+        // convert to email strings to type safe objects
+        .map(_.map { case (userEmail, saEmail) => (RawlsUserEmail(userEmail), BondServiceAccountEmail(saEmail))})
+        // group results to the right return value shape
+        .map(x => CollectionUtils.groupByTuples(x))
     }
   }
 
   private def existingRecordsForUserQuery(workspaceName: WorkspaceName, userEmail: RawlsUserEmail): Query[WorkspaceRequesterPaysTable, WorkspaceRequesterPaysRecord, Seq] = {
     val workspaceSubquery = workspaceQuery.filter(ws => ws.namespace === workspaceName.namespace && ws.name === workspaceName.name).map(_.id)
     workspaceRequesterPaysQuery.filter(_.workspaceId in workspaceSubquery).filter(_.userEmail === userEmail.value)
   }
+
+  private def existingRecordsForWorkspaceQuery(workspaceName: WorkspaceName): Query[WorkspaceRequesterPaysTable, WorkspaceRequesterPaysRecord, Seq] = {
+    for {
+      ws <- workspaceQuery if ws.namespace === workspaceName.namespace && ws.name === workspaceName.name
+      rp <- workspaceRequesterPaysQuery if rp.workspaceId === ws.id
```

Comment:
the syntax throws me off a bit, especially with ws and rp not being defined :/
do you mind walking me through it (either here or on a quick chat)

Reply:
this is how we write db queries in slick. This boils down to
```
select rp.*
from WORKSPACE ws
join WORKSPACE_REQUESTER_PAYS rp on rp.workspace_id = ws.id
where ws.namespace = ? and ws.name = ?
```

Reply:
I see! Thanks for the explanation. I probably need to get used to it.